handler/datasource/prepare.go

```diff
@@ -0,0 +1,293 @@
+		return err
+	}
+
+	return PrepareSource(ctx, db, datasourceHandlerResolver, source, false)
+}
+func markPackJobsReady(
+	ctx context.Context,
+	db *gorm.DB,
+	sourceID uint32,
+) error {
+	return database.DoRetry(ctx, func() error {
+		return db.Model(&model.PackJob{}).Where("source_id = ? AND packing_state = ?", sourceID, model.Created).Update("packing_state", model.Ready).Error
+	})
+}
+
+func updatePackJob(
+	ctx context.Context,
+	db *gorm.DB,
+	packJobID uint32,
+	state model.WorkState,
+	fileRangeIDs []uint64,
+) error {
+	return database.DoRetry(ctx, func() error {
+		return db.Transaction(
+			func(db *gorm.DB) error {
+				err := db.Model(&model.PackJob{}).Where("id = ?", packJobID).Update("packing_state", state).Error
+				if err != nil {
+					return fmt.Errorf("failed to update pack job: %w", err)
+				}
+				fileRangeIDChunks := util.ChunkSlice(fileRangeIDs, util.BatchSize)
+				for _, fileRangeChunk := range fileRangeIDChunks {
+					err = db.Model(&model.FileRange{}).
+						Where("id IN ?", fileRangeChunk).Update("pack_job_id", packJobID).Error
+					if err != nil {
+						return fmt.Errorf("failed to update items: %w", err)
+					}
+				}
+				return nil
+			},
+		)
+	})
+}
+
+type fileRangeSet struct {
+	fileRanges []model.FileRange
+	carSize    int64
+}
+
+const carHeaderSize = 59
+
+func newFileRangeSet() *fileRangeSet {
+	return &fileRangeSet{
+		fileRanges: make([]model.FileRange, 0),
+		// Some buffer for header
+		carSize: carHeaderSize,
+	}
+}
+
+func (frs *fileRangeSet) add(fileRanges []model.FileRange) {
+	frs.fileRanges = append(frs.fileRanges, fileRanges...)
+	for _, fileRanges := range fileRanges {
+		frs.carSize += toCarSize(fileRanges.Length)
+	}
+}
+
+func (frs *fileRangeSet) addIfFits(fileRange model.FileRange, maxSize int64) bool {
+	nextSize := toCarSize(fileRange.Length)
+	if frs.carSize+nextSize > maxSize {
+		return false
+	}
+	frs.fileRanges = append(frs.fileRanges, fileRange)
+	frs.carSize += nextSize
+	return true
+}
+
+func (frs *fileRangeSet) reset() {
+	frs.fileRanges = make([]model.FileRange, 0)
+	frs.carSize = carHeaderSize
+}
+
+func (frs *fileRangeSet) fileRangeIDs() []uint64 {
+	return underscore.Map(frs.fileRanges, func(fileRange model.FileRange) uint64 {
+		return fileRange.ID
+	})
+}
+
+func toCarSize(size int64) int64 {
+	out := size
+	nBlocks := size / 1024 / 1024
+	if size%(1024*1024) != 0 {
+		nBlocks++
+	}
+
+	// For each block, we need to add the bytes for the CID as well as varint
+	out += nBlocks * (36 + 9)
+
+	// For every 256 blocks, we need to add another block.
+	// The block stores up to 256 CIDs and integers, estimate it to be 12kb
+	if nBlocks > 1 {
+		out += (((nBlocks - 1) / 256) + 1) * 12000
+	}
+
+	return out
+}
+
+func nextAvailablePackJob(
+	db *gorm.DB,
+	sourceID uint32,
+) (*model.PackJob, error) {
+	var packJob model.PackJob
+	err := db.Where(model.PackJob{SourceID: sourceID, PackingState: model.Created}).Preload("FileRanges").FirstOrCreate(&packJob).Error
```

Comment:
FirstOrCreate needs to be protected by a transaction and have a database.DoRetry wrapper around it.